net/interfaces/defaultroute_ios.go

```diff
@@ -0,0 +1,128 @@
+	// seems to set the delegate interface only once, upon the *creation* of the VPN tunnel.
+	// If a network transition (e.g. from Wi-Fi to Cellular) happens while the tunnel is
+	// connected, it will be ignored and we will still try to set Wi-Fi as the default route
+	// because the delegated interface is not updated by the NetworkExtension framework.
+	//
+	// We work around this on the Swift side with a NWPathMonitor instance that observes
+	// the interface name of the first currently satisfied network path. Our Swift code will
+	// call into `UpdateLastKnownDefaultRouteInterface`, so we can rely on that when it is set.
+	//
+	// If for any reason the Swift machinery didn't work and we don't get any updates, here
+	// we also have some fallback logic: we try finding a hardcoded Wi-Fi interface called en0.
+	// If en0 is down, we fall back to cellular (pdp_ip0) as a last resort. This doesn't handle
+	// all edge cases like USB-Ethernet adapters or multiple Ethernet interfaces, but is good
+	// enough to ensure connectivity isn't broken.
+
+	// Start by getting all available interfaces.
+	interfaces, err := netInterfaces()
+	if err != nil {
+		log.Printf("defaultroute_ios: could not get interfaces: %v", err)
+		return d, ErrNoGatewayIndexFound
+	}
+
+	// Did Swift set lastKnownDefaultRouteInterface? If so, we should use it and don't bother
+	// with anything else. However, for sanity, do check whether Swift gave us with an interface
+	// that exists, is up, and has an address.
+	if swiftIfName := lastKnownDefaultRouteIfName.Load(); swiftIfName != "" {
+		for _, ifc := range interfaces {
+			if ifc.Name != swiftIfName {
+				continue
+			}
+
+			if !ifc.IsUp() {
+				log.Printf("defaultroute_ios: interface %s provided by Swift isn't up", ifc.Name)
+				break
+			}
+
+			addrs, _ := ifc.Addrs()
+			if len(addrs) == 0 {
+				log.Printf("defaultroute_ios: interface %s provided by Swift has no addresses", ifc.Name)
+				break
+			}
+
+			log.Printf("defaultroute_ios: returning interface %s provided by Swift", ifc.Name)
+			d.InterfaceName = ifc.Name
+			d.InterfaceIndex = ifc.Index
+			return d, nil
+		}
+	}
+
+	// Start of our fallback logic if Swift didn't give us an interface name, or gave us an invalid
+	// one.
+	// We start by attempting to use the Wi-Fi interface, which on iPhone is always called en0.
+	for _, ifc := range interfaces {
+		if ifc.Name != "en0" {
+			continue
+		}
+
+		if !ifc.IsUp() {
+			log.Println("defaultroute_ios: Wi-Fi interface en0 isn't up, will try cellular")
+			break
+		}
+
+		addrs, _ := ifc.Addrs()
+		if len(addrs) == 0 {
+			log.Println("defaultroute_ios: Wi-Fi interface en0 has no addresses")
+			break
+		}
+
+		log.Println("defaultroute_ios: returning Wi-Fi interface en0")
+		d.InterfaceName = ifc.Name
+		d.InterfaceIndex = ifc.Index
+		return d, nil
+	}
+
+	// Did it not work? Let's try with Cellular (pdp_ip0).
+	for _, ifc := range interfaces {
```

Comment:
This loop is very similar to the two other loops before; perhaps it's worth abstracting away into a separate function?

Reply:
I had written a similar comment, but... +1 to this. If it's any easier, you could make it a closure right below the call to `netInterfaces`; something like:

```go
	interfaces, err := netInterfaces()
	if err != nil {
		log.Printf("defaultroute_ios: could not get interfaces: %v", err)
		return d, ErrNoGatewayIndexFound
	}

	getInterfaceByName := func(name string) Interface {
		for _, ifc := range interfaces {
			if ifc.Name != name {
				continue
			}

			if !ifc.IsUp() {
				log.Println("defaultroute_ios: interface %s isn't up", name)
				return nil
			}

			addrs, _ := ifc.Addrs()
			if len(addrs) == 0 {
				log.Println("defaultroute_ios: interface %s has no addresses", name)
				return nil
			}
			return ifc
		}
		return nil
	}
```

(or something, this is just an example 😅)

Reply:
Simplied, this is much better, thanks!